### DISCUSSION TITLE
Using something different than bytes array for EVP cipher API (EVP_SKEY)

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#21355] [General Discussion]
I have an impression that we already are able to use opaque structure in EVP cipher API. 

We have the `EVP_CipherInit_ex2` API function accepting params[] array and passing the array together with key and iv to the provider.

It means that we can use a parameter of type OSSL_PARAM_UTF8_STRING for, say, PKCS#11 URI or CKA_LABEL, or OSSL_PARAM_UNSIGNED_INTEGER for key handle. As the implementation is provider-specific, it will be suitable for a provider author to implement the cipher initialization from this parameter.

What I don't understand is the necessary level of abstraction. Opaque blobs are not interoperable between providers, non-extractable keys are not serializable so it looks like provider-specific format and provider-specific parameters are enough.

**Comment by paulidale** [comment]
I'm not sure opaque structures should be passed via params.  If we did, it would be an octet string but I do feel it is a bad idea.
What use case requires passing opaque structures around?

We have discussed a symmetric key type in the past -- it's not going to be quite the same.

**Comment by beldmit** [comment]
For pkcs11 purposes we don't really need opaque structs - URIs will be enough. I believe that for other cases (TPM) we also have some identifier.

**Comment by beldmit** [comment]
@paulidale I adjusted the title

**Comment by paulidale** [comment]
Ah, okay.  Symmetric key type.  Makes sense.

**Comment by mattcaswell** [comment]
If we're talking about a symmetric key type then we have long talked about having an EVP_SKEY. It would probably require a key manager in much the same way that EVP_PKEY does and I imagine it would work in a very similar way to EVP_PKEY.

**Comment by beldmit** [comment]
I'm not sure if we really need this object. We either have some provider-specific type or EVP_CIPHER_CTX object already having it set.

**Comment by mattcaswell** [comment]
I expect you might want to call "load" or "generate" type calls and get back a handle on an EVP_SKEY which you then subsequently pass to an EVP_CIpher* call.

**Comment by beldmit** [comment]
It also can be dealt via get_parameters, I think?

**Comment by mattcaswell** [comment]
Yuck. I really don't think we want to do that. Why would we not have it as a first-class object? That seems very limiting. I imagine we will want to do much of the same sort of stuff that we do with EVP_PKEY with EVP_SKEY as well. E.g. exchange a new key using some kex algo in one provider and then seamlessly export/import it into a different provider for actual use.

**Reply by t8m** [reply]
Or even exchange and then use it within one provider not leaving the provider boundary at all.

**Reply by beldmit** [reply]
Yes, it makes sense to me, so let me think about it some more time

**Comment by beldmit** [comment]
IRL the asymmetric key is either exportable to the necessary extent to be importable in the different provider or is a provider-scoped (`default` and `fips` providers can be considered the same for this purpose). We had plenty bugs related to this import-export dance already, I think.

For symmetric key there is one natural representation for data exchange - bytes array + length - and some provider-specific ones.

**Reply by levitte** [reply]
For symmetric keys, my first thought was that the export/import dance would be quite easy, considering that an exported symmetric key should be "a bunch of bytes" and that's it.  But knowing us, I can imagine that we'd want to attach all sorts of attributes to an EVP_SKEY, and we might end up making it more complicated.  [sigh]

I have been ruminating an idea for a while, that might allow us to get away from the export/import dance: attaching primitive operations to whatever provider side mgmt op we decide on (keymgmt would actually be viable in my view).  As a matter of fact, we kinda sorta already have that with the generator functions, but we could generalize that a bit differently.

**Comment by mattcaswell** [comment]
> For symmetric key there is one natural representation for data exchange - bytes array + length 

By doing it this way the import/export is the application's responsibility. With EVP_SKEY it is OpenSSL's  and is entirely invisible to the application.

If you make it the application's responsibility then the application has to do the right thing with an array of bytes (e.g. remember to cleanse it afterwards etc). By using an EVP_SKEY the application just has an abstract "handle" on the key. The key might never leave the provider at all (as mentioned by @t8m above).

**Comment by beldmit** [comment]
Thank you, I will reconsider it

**Comment by slontis** [comment]
I think EVP_SKEY is the correct thing to do here.. The issue is that it requires new API's that wont be available to a major release.

**Reply by paulidale** [reply]
New APIs can be added to any release, so we could introduce this for 3.3.

**Reply by t8m** [reply]
Of course this won't be supported by existing providers. I assume there could be some compat layer added to libcrypto that would do some export/import dance with providers that won't support direct use of EVP_SKEYs.

**Comment by simo5** [comment]
Any reason why reusing a EVP_PKEY to carry symmetric keys as well would be really bad?

I am ok with the EVP_SKEY idea, just wondering why there is a need to fragment the "key handling" concept. After all Public Keys are carried via EVP_PKEY as well, so this type already carries 2 key types...

**Reply by t8m** [reply]
IMO EVP_PKEYs and EVP_SKEYs will be inherently  used in different set of operations and operational API calls. Yes, there might be some commonality in terms of keymgmt APIs but other than that I think it is a good thing to have type separation for different types of objects.

**Reply by paulidale** [reply]
EVP_PKEYs are already used for a variety of operations.  I see some sense in adding symmetric crypto to the mix.

**Comment by slontis** [comment]
EVP_PKEY has already been hacked to do other things.. It may be easier to just keep adding to EVP_PKEY, but I dont think it is a 'clean' way moving forward. It assume it was intended to represent a keypair.

**Reply by beldmit** [reply]
Well, for MAC-as-signature (old interface) it was not a keypair

**Comment by slontis** [comment]
And the MAC is a very good example of why extending EVP_PKEY is the wrong thing to do.

**Comment by beldmit** [comment]
OK, what about this approach?

We already have the function
```
EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                   const unsigned char *key, const unsigned char *iv,
                   int enc, const OSSL_PARAM params[]);
```
                            
To find out the parameters supported by a particular EVP_CIPHER object,
we introduce the function

```
const OSSL_PARAM *EVP_CIPHER_settable(const EVP_CIPHER *cipher);
```
       
similar to `EVP_PKEY_fromdata_settable`

To get/set the currently set key as a EVP_SKEY object, we introduce the API
       
```
EVP_SKEY * EVP_CIPHER_CTX_get_EVP_SKEY(const EVP_CIPHER_CTX *ctx);
int EVP_CIPHER_CTX_get_EVP_SKEY(const EVP_CIPHER_CTX *ctx, const EVP_SKEY *skey);
```
The getter should create a new EVP_SKEY object to be freed by `EVP_SKEY_free` function.

To get the currently set parameters, we use
```
int EVP_SKEY_todata(const EVP_SKEY *key, OSSL_PARAM **params);
int EVP_SKEY_export(const EVP_SKEY *key, OSSL_PARAM **params, OSSL_CALLBACK *export_cb, void *export_cbarg);
```

Duplicating the EVP_SKEY object
```
EVP_SKEY * EVP_SKEY_dup(const EVP_SKEY * skey);
``` 

Free function:
```
void EVP_SKEY_free(EVP_SKEY *key);
```

**Reply by paulidale** [reply]
We've got `EVP_CIPHER_settablele_params()` which seems more appropriate than `EVP_CIPHER_settable()`.
I'm with @slontis, a new cipher init call is the way to go.

**Reply by simo5** [reply]
I do not see how this works with Key Exchange/Key Derivation operations that produce bytes today, and instead need to return an EVP_SKEY. If you new APIs need to be created for those, might as well create new CiPherInit API that explicitly takes an EVP_SKEY ?

**Comment by slontis** [comment]
Would we need a EVP_CipherInit call that takes a EVP_SKEY..
For PKCS11 I imagine this internally would be just an id rather than bytes..

**Reply by beldmit** [reply]
For PKCS11 it may be either a handle or a combination of parameters to get a handle.

**Reply by t8m** [reply]
We would need it as EVP_SKEYs could be produced not only by fromdata but also as result from EVP_KDF operations.

**Reply by t8m** [reply]
And/or perhaps also a result of EVP_PKEY_derive().

**Reply by simo5** [reply]
Yes, any operation that produces a bunch of bytes usable as a key should use the type in theory.

**Comment by simo5** [comment]
Until we have something like EVP_SKEY, I wonder how really bad it would be if I returned a pkcs11 uri as the "bunch of bytes".

**Reply by beldmit** [reply]
Quite bad - the key has a fixed length.

**Comment by beldmit** [comment]
Creation

```
EVP_SKEY * EVP_SKEY_new(void)
```

setting parameters

```
int EVP_SKEY_fromdata(EVP_CIPHER_CTX *ctx, EVP_SKEY **pskey, OSSL_PARAM params[]);
int EVP_SKEY_assign(EVP_CIPHER_CTX *ctx, EVP_SKEY *pskey, OSSL_PARAM params[]);
```

Derivation:

```
int EVP_PKEY_derive_SKEY(EVP_PKEY_CTX *ctx, EVP_SKEY *pskey);
```
similar to `EVP_PKEY_derive`

We already have the function

```
EVP_CipherInit_ex2(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                   const unsigned char *key, const unsigned char *iv,
                   int enc, const OSSL_PARAM params[]);
```

and can introduce

```
EVP_CipherInit_ex3(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
                   const EVP_SKEY *skey, const unsigned char *iv,
                   int enc, const OSSL_PARAM params[]);
```

To find out the parameters supported by a particular EVP_CIPHER object,
we introduce the function

```
const OSSL_PARAM *EVP_CIPHER_settable(EVP_CIPHER *cipher);
```

similar to `EVP_PKEY_fromdata_settable`

To get/set the currently set key as a EVP_SKEY object, we introduce the API
```
EVP_SKEY * EVP_CIPHER_CTX_get_EVP_SKEY(const EVP_CIPHER_CTX *ctx);
int EVP_CIPHER_CTX_set_EVP_SKEY(const EVP_CIPHER_CTX *ctx, const EVP_SKEY *skey);
```
The getter should create a new EVP_SKEY object to be freed by `EVP_SKEY_free` function.

To get the currently set parameters, we use
```
int EVP_SKEY_todata(const EVP_SKEY *key, OSSL_PARAM **params);
int EVP_SKEY_export(const EVP_SKEY *key, OSSL_PARAM **params, OSSL_CALLBACK *export_cb, void *export_cbarg);
```

Duplicating the EVP_SKEY object
```
EVP_SKEY * EVP_SKEY_dup(const EVP_SKEY * skey);
```

Free function:
```
void EVP_SKEY_free(EVP_SKEY *key);
```

**Comment by t8m** [comment]
We could avoid adding `EVP_CipherInit_ex3()`. You could call  EVP_CIPHER_CTX_set_EVP_SKEY() and then EVP_CipherInit_ex2() with NULL key.

I do not think we need EVP_CIPHER_CTX_get_EVP_SKEY() - IMO especially we do not want "silent" creation of new EVP_SKEY object by this function.

EVP_SKEY_assign() seems also unnecessary to me - we want to avoid the mistake with mutable EVP_PKEY objects IMO.

EVP_CIPHER_settable() is a wrong name - it should be EVP_SKEY_fromdata_settable() instead.

**Reply by beldmit** [reply]
Fixed in the design document draft.

**Comment by t8m** [comment]
It should be possible to produce EVP_SKEY with EVP_KDF_derive - i.e., we need EVP_KDF_derive_skey. But we also need EVP_KDF_set_skey() and EVP_MAC_set_skey(). And there are also encapsulate/decapsulate functions which should be possible to use with transferring keys.

**Reply by beldmit** [reply]
Provided functions for KDF/MAC, not sure about encapsulate/decapsulate yet.

**Comment by t-j-h** [comment]
I'd be more in favour of an EVP_CipherInit_skey (or whatever we call it) than using generic ex3 or reusing the existing ex2 with a null key bytes approach which is a path that leads to a lot more questions and mistakes. 

The null-abuse can be reduced by having a special pointer than means using an already-set EVP_SKEY - but that still would be messy.

EVP_SKEY needs to have a lifecycle like EVP_PKEY and not just be seen as a pure bytes-of-a-key container. The concept will be to get to the point of being able to have PKCS#11 handling that works in a more direct manner.

**Reply by t8m** [reply]
The pattern of passing NULL key bytes to EVP_CipherInit is already supported as you can use that pattern to re-init the context with the same key but changed IV.

Another question is whether the EVP_SKEY should have an application accessible key type (algorithm) or if it should be just purely opaque and the provider would be responsible for any type mismatch handling - i.e. a different length of a key generated with EVP_KDF_derive() than needed for a particular cipher.

**Reply by beldmit** [reply]
I agree with @t8m here, we already allow setting key/iv separately and we don't need either _ex3 or a pecial pointer value

**Reply by simo5** [reply]
It would be more ergonomic to be able to set up all of it with a single init call, especially if this will ever become the recommended API to use.

**Reply by beldmit** [reply]
OK, fixed

**Comment by levitte** [comment]
I'm a little lost at what EVP_SKEY is meant to accomplish here.  Apart from the possibility to catch size errors before initializing a Cipher operation, and possibly being a reference to a hardware-backed key, what's the conceptual difference between an EVP_SKEY and "a bunch of bytes"?

Looking at #21573, I'm strongly missing the conceptual picture (what it's supposed to represent, what it accomplishes, why have it at all...)

**Reply by beldmit** [reply]
An opaque object provides us some flexibility, but mainly it is desired to cover the PKCS#11 and probably TPM keys which have a handle

**Comment by slontis** [comment]
@levitte This is motivated by P11, in which case a EVP_SKEY is more likely to be represented by an id..

**Comment by levitte** [comment]
Yeah, and so there should probably be a loading function of some sort, to "load" the key by id, like there already is in keymgmt, and if we're to "load" it from a container of some sort, shouldn't there STORE interface be amended to accommodate that?  That would, BTW, answer the PKCS#11 URI discussion that I see higher up.

~~Something one might want to think about as well in connection with this is [drum-roll] encoding (for extractable keys only, of course).  I was just now looking at #19340, and was thinking that it would be really cool if it was possible for a provider to not only provide a new KDF, but also to provide a way to encode the resulting hashed password in a convenient way for /etc/shadow, without having to hack in that knowledge into apps/passwd.c...~~
(UPDATE: that section was beside the point, please ignore it)

**Comment by levitte** [comment]
One thing that isn't clear in these discussions (or in #21573) is, do EVP_SKEYs have a key type?  EVP_PKEYs do, and one can thereby say that they are attached to a set of operations (and conceptually speaking, a set of primitives), implemented a certain way.

Another way to see this is if they are really seen as representing "a bunch of bytes", or if they ultimately represent a key schedule for an algorithm.  Conceptually speaking, this is what `keymgmt` does, so if EVP_SKEY is just meant to represent "a bunch of bytes", then `keymgmt` isn't really suitable for it...

**Reply by beldmit** [reply]
I don't know. I think the algorithm is specified in PKCS#11 tokens. I don't know what happens inside TPMs.

**Reply by levitte** [reply]
For PKCS#11, I assume you're talking about CKA_ALLOWED_MECHANISMS?  Spending some time trying to read the spec, that's what I found...

**Comment by simo5** [comment]
PKCS#11 has a key type argument, for secret (symmetric) keys thisis what you genrally see used in object attributes:
```
CK_OBJECT_CLASS class = CKO_SECRET_KEY;
CK_KEY_TYPE keyType = CKK_GENERIC_SECRET;
```

However there are also symmetric key types for specific operations: CKK_DES, CKK_AES, etc..

See `6.10.2 AES secret key objects` in v3.1 spec as an example.

So yeah, pkcs#11 does have a concept of a typed symmetric key.

HTH

**Reply by simo5** [reply]
Btw there are also keys for HMAC operations like CKK_SHA256_HMAC etc... check the pkcs11t.h header for the list of existing definitions of key types, they all start with CKK_

**Reply by t8m** [reply]
That's why I suggest having both possibilities - a typed and untyped EVP_SKEY. However I am not even sure the type needs to live outside the provider that holds the key. If the key is exported then why we would need to keep it typed? I still expect the exported key representation to be just a bunch of seemingly random bytes.

**Reply by levitte** [reply]
So what I'm hearing from you, @t8m, is that an EVP_SKEY could essentially be a combination of these:

- A reference to a provider-backed symmetric key
- An array of bytes

If no provider-backed symmetric key has been loaded, that reference will of course be NULL.
If a provider-backed symmetric key has been loaded, it may be a source for populating the array of bytes.
The array of bytes, once populated, can be used by *all other* providers.

That's actually a good point, there isn't really a need to have a per-provider cache, 'cause once the symmetric key is exported, that export should be usable by everyone, only subject to size limits.

**Reply by t8m** [reply]
Yeah, the array of bytes would not really be a per-provider cache, just a different representation of the same key.

**Reply by simo5** [reply]
I think the only challenge for the pkcs#11 provider is to know how to type the key on import.
But we can probably deal with that by creating a session key and re-importing it with the right type if needed.

Note that we are still missing a good way to import keys permanently into tokens (as opposed to ephemeral imports into session objects). This is something that is a bit at odds with how openssl is structured, given the store API only deals with loading stuff but not saving it, and the saving is instead deferred to encoding into some bunch of bits and then taking those bits and shoving them into a file.

This obviously doesn't work for keys to be imported into a token. I am not sure we have to care about it in order to introduce the EVP_SKEY concept, but just bringing it up for completeness.

**Reply by t8m** [reply]
IMO that needs a public STORE management API - I do not think we need to resolve it with introduction of EVP_SKEY. Any solution for that needs to work also for EVP_PKEYs so I assume it could be added later for both.

**Reply by levitte** [reply]
**Side note**:

The STORE API can certainly be extended to save as well.  That wasn't the original scope of that API, but it was built to allow that sort of additional functionality when needed.  Like, we now have functionality to delete STORE objects!

Of course, a save API will require some thinking and deliberation...

**Comment by beldmit** [comment]
I think this can be closed now as we have a basic implementation already.
